design/external-compaction.md

```diff
@@ -0,0 +1,249 @@
+
+## APIs
+
+### Compaction Coordinator RPC API
+
+```
+/**
+ * Called by tablet servers to get the status of compaction
+**/
+ExternalCompactionStatus[] getCompactionStatus(TabletId[])
+
+
+/**
+ * Called by tablet servers to get the number of running and queued compaction requests
+**/
+ExternalCompactionStats getCompactionStats(TabletId, String queueName)
+
+
+/**
+ * Called by compactor processes to get work to do for a queue.  The compactor Id can be used    to check if the process is alive. The implementation will find the top tablet in its queue and then    call reserveJob for that tablet on the tablet server.  This could be a blocking RPC call where it does not return until there is work, however not sure how well this will work.
+ */
+CompactionJob getCompactionJob(String queueName, CompactorId)
+
+
+/**
+ * Called by compactor process to indicate work is done. The implementation will call  finishedJob on the tablet servers.
+ **/
+void finishedJob(CompactionJob, CompactionStats)
+
+
+/**
+  * Called by compactor process to update the coordinator with the state of the running compaction
+ **/
+void updateCompactionStatus(CompactionJob, State, message, timestamp)
+```
+
+### Compactor RPC API
+
+```
+/**
+  * Called by CompactionCoordinator to cancel the compaction work
+**/
+void cancel(CompactionJob)
+```
+
+### Tablet Server Compaction RPC API
+
+```
+/**
+  Summary information about a tablet servers external compaction queue.
+ */
+Class TabletServerCompactionQueueSummary {
+  // the name of an external queue for which information is summarized
+  Sting queueName;
+   // the priority for which this information is summarized
+   Long priority;
+   // the number of tablets queued on a tserver for a given queue and priority
+   Int queuedCount;
+}
+
+
+/**
+ * Called by the coordinator to get summary information about all of a tablets servers external compaction queues.
+ */
+List<TabletServerCompactionQueueSummary> getTabletServerQueueInfo()
+
+
+/**
+ * Called to atomically reserve compaction work for a tablet on a given queue with a priority level >= than what was passed.  Will return null if there is no work meeting criteria.  Tablets will treat these files reserved as part of a running compaction for planning future compactions. Will use CompactorId to check if process running compactions is alive.  Information about the reservation will be persisted by the tablet in the metadata table so that if the tablet loads elsewhere it will not lose track of the external compaction.
+ */
+CompactionJob reserveJob(String queueName, long priority, CompactorId)
```

Comment:
Would the priority be system generated (like age + size)?

Reply:
The default planner currenty uses the compaction type and #of files to construct a priority.   The following are links to the code that currently creates the priority.

https://github.com/apache/accumulo/blob/b12d1103d788473168616f28ae87a65bb76aa880/core/src/main/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlanner.java#L236

https://github.com/apache/accumulo/blob/b12d1103d788473168616f28ae87a65bb76aa880/core/src/main/java/org/apache/accumulo/core/util/compaction/CompactionJobPrioritizer.java#L33

@dlmarion  and I were talking and concluded that long is probably too high cardinality for priority.  Maybe a 16 bit or 8 bit integer would be better.   Then maybe the prio could be type:log2(numFiles)